### PR TITLE
Keep image processor backends in sync on keys and dtypes

### LINKED ISSUE #48738: Image processor backends return different keys and dtypes for the same input
### System Info

- transformers: `main` at `df04b01222`
- PyTorch 2.11.0, torchvision 0.26.0, Python 3.13.15, Linux, CPU
- The reproduction builds the processors from their class defaults and downloads nothing.

### Who can help?

@molbap @guarin

### Reproduction

I swept all 92 model families that ship both an `XImageProcessor` (torchvision) and an `XImageProcessorPil`, ran the same image through both, and compared every returned key. Two kinds of divergence came out.

**1. `pixel_attention_mask` is float32 on the torchvision backend, int64 on the PIL one** (Idefics2, Idefics3, SmolVLM):

```python
import numpy as np, torch
from PIL import Image
import transformers

img = Image.fromarray(np.random.RandomState(0).randint(0, 255, (480, 640, 3), dtype=np.uint8))
for name in ["Idefics2ImageProcessor", "Idefics3ImageProcessor", "SmolVLMImageProcessor"]:
    a = getattr(transformers, name)()(img, return_tensors="pt")["pixel_attention_mask"]
    b = getattr(transformers, name + "Pil")()(img, return_tensors="pt")["pixel_attention_mask"]
    print(name, a.dtype, b.dtype, torch.equal(a.long(), b.long()))
```

```text
Idefics2ImageProcessor torch.float32 torch.int64 True
Idefics3ImageProcessor torch.float32 torch.int64 True
SmolVLMImageProcessor torch.float32 torch.int64 True
```

The values agree, only the dtype differs. `Idefics3Model.forward` documents the argument as `pixel_attention_mask (torch.LongTensor, optional)`.

**2. Fuyu returns a different set of keys per backend:**

```python
img = Image.fromarray(np.random.RandomState(0).randint(0, 255, (480, 640, 3), dtype=np.uint8))
print(sorted(transformers.FuyuImageProcessor()(img, return_tensors="pt").keys()))
print(sorted(transformers.FuyuImageProcessorPil()(img, return_tensors="pt").keys()))
```

```text
['image_scale_factors', 'image_sizes', 'image_unpadded_heights', 'image_unpadded_widths', 'images']
['image_scale_factors', 'image_unpadded_heights', 'image_unpadded_widths', 'images']
```

`FuyuProcessor.unused_input_names` lists `image_sizes`, and `FuyuProcessor` drops it before the model, so this one only bites when the image processor is used directly. The PIL backend already computes the value, it just never puts it in the output.

### Expected behavior

Both backends produce the same keys with the same dtypes, so that swapping backends (which happens automatically when torchvision is not installed, see #48616) does not change the output contract.

### Reason and Possible fixes

For the masks, the per-image mask that `pad()` builds is explicitly `torch.int64`, but the batch buffer it is written into is allocated without a dtype:

```python
pixel_attention_masks = torch.zeros(
    len(processed_images),
    max_num_images,
    *(max_height, max_width),
    device=device,
)
```

so the assignment casts the mask to the default float dtype. Every other image processor that builds a mask passes an integral dtype explicitly: DETR, Conditional DETR, Deformable DETR, Grounding DINO, RT-DETR, RF-DETR, YOLOS, Mask2Former and MaskFormer use `dtype=torch.int64`, Aria uses `torch.bool`, Mllama uses `torch.long`. Only these three buffers omit it.

This got past CI because `test_backends_equivalence` compares `pixel_values` and nothing else, and the Idefics2, Idefics3 and SmolVLM overrides of that test compare the masks as `pixel_attention_mask.float()`, which normalizes the dtype away before the assertion.

I would like to send a PR that:

1. passes `dtype=torch.int64` in the three buffers (SmolVLM regenerated through `modular_smolvlm.py`),
2. returns `image_sizes` from the Fuyu PIL backend,
3. extends the shared `test_backends_equivalence` and `test_backends_equivalence_batched` to compare every returned key (same keys, same dtype and shape, exact equality for integral outputs, existing tolerances for float ones) and drops the `.float()` calls in the three overrides.

I have that ready. With the stronger test but without the fixes, 7 of the 8 backend tests for these four models fail; with the fixes all of them pass, and the whole `tests/models/*/test_image_processing_*.py -k backends_equivalence` suite is 190 passed, 63 skipped, so no other model is affected. Happy to split it or narrow the test change if you prefer.

Searched for existing reports before filing: `pixel_attention_mask dtype`, `pixel_attention_mask float`, `backends_equivalence`, `ImageProcessorPil`, `image_sizes fuyu`, and open PRs touching these files. The only overlap is the draft #48192, which renames loop variables in the same functions but does not touch the mask buffers or the Fuyu output dict.

AI assistance: I used Claude Code to run the backend sweep across all 92 processor pairs and to draft this report. I verified every result above myself and reviewed the change end to end.

### Checklist

- [x] I have checked that there is no similar issue in the repo (**required**)
- [x] I have read the [documentation](https://github.com/huggingface/transformers/tree/main/docs) (**required**)
- [x] I have provided a minimal working example to reproduce the bug (**required**)

### PR DESCRIPTION
**Pull request by yupengtang**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48739) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48739)

# What does this PR do?

Fixes #48738.

The torchvision backends of Idefics2, Idefics3 and SmolVLM allocate the batch `pixel_attention_mask` buffer without a dtype:

```python
pixel_attention_masks = torch.zeros(
    len(processed_images),
    max_num_images,
    *(max_height, max_width),
    device=device,
)
```

so the int64 masks that `pad()` builds are stored as float32, while the PIL backend returns int64 and `Idefics3Model.forward` documents the argument as a `LongTensor`. Every other image processor that builds a mask passes an integral dtype explicitly (DETR and friends, Mask2Former, MaskFormer and YOLOS use `torch.int64`, Aria uses `torch.bool`, Mllama uses `torch.long`), so this is just a missing argument in three places. SmolVLM is regenerated through `modular_smolvlm.py`.

The Fuyu PIL backend already computes `original_image_sizes` but never returns it, so its output was missing the `image_sizes` key that the torchvision backend returns.

`test_backends_equivalence` compared `pixel_values` and nothing else, and the Idefics2, Idefics3 and SmolVLM overrides compared the masks as `pixel_attention_mask.float()`, which normalizes the dtype away before the assertion. This PR compares every returned key instead: same keys, same dtype and shape, exact equality for integral outputs and the existing tolerances for float ones. The `.float()` calls are no longer needed.

## Testing

```
pytest tests/models/{idefics2,idefics3,smolvlm,fuyu}/test_image_processing_*.py -k backends_equivalence
```

- with the stronger test but without the source fixes: 7 failed, 1 passed (the remaining pass is the Fuyu batched test, which compares only `images`)
- with this branch: those four test files run in full, 95 passed, 17 skipped

```
pytest tests/models/*/test_image_processing_*.py -k backends_equivalence
```

- 190 passed, 63 skipped, no failures, so no other model is affected by the stricter comparison

Also ran `ruff check` and `ruff format --check` (0.14.10, the version pinned in `setup.py`) and `utils/check_modular_conversion.py --files src/transformers/models/smolvlm/modular_smolvlm.py` on the changed files, all clean. Python 3.13, torch 2.11.0, torchvision 0.26.0, CPU.

The draft #48192 touches the same functions but renames loop variables rather than the mask buffers, so the two should not collide.

## AI assistance

I used Claude Code to sweep the 92 processor pairs and to help draft the change. I verified each divergence myself, wrote and ran the before and after test runs, and reviewed every line of the diff.

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://huggingface.co/docs/transformers/contributing) and the
      [Pull Request](https://huggingface.co/docs/transformers/pr_checks) checks?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. (#48738)
- [ ] Did you make sure to update the documentation with your changes according to the [guidelines](https://github.com/huggingface/transformers/tree/main/docs)?
- [x] Did you write any new necessary [tests](https://huggingface.co/docs/transformers/testing)?

## Who can review?

@molbap @guarin

### CODE REVIEW

**Comment by yupengtang** [comment]
Both red shards look unrelated to this change:

- `tests_processors [shard 3/8]`: 854 passed, and the one error is `httpx.RemoteProtocolError` in `Idefics3ProcessorTest::test_kwargs_overrides_default_subprocessor_kwargs_2_audio`, a hub download that dropped mid-request.
- `tests_generate [shard 3/8]`: 1192 passed, and the failure is `Kosmos2_5ModelTest::test_cached_decode_matches_cacheless`. That one passes for me locally on this branch, and the diff only touches four image processors plus the image processing tests, so it does not reach the generation path.

I cannot rerun the jobs myself; happy to push a rebase if you want a clean run.

File: tests/test_image_processing_common.py [hunk 1]
```diff
@@ -282,6 +282,43 @@ def _assert_tensors_equivalence(self, tensor1, tensor2, atol=1e-1, rtol=1e-3, me
...
+
+        Float outputs are compared with tolerances because the backends resize differently;
+        discrete outputs (masks, sizes, token ids) must match exactly.
+        """
+        self.assertEqual(
+            set(reference_encoding.keys()),
+            set(encoding.keys()),
+            f"{backend_name} returns different keys than {reference_backend}",
+        )
+        for key in reference_encoding:
```
Comment (zucchini-nlp) [inline tests/test_image_processing_common.py]: niiice, it should have been checking all keys from the beginning

File: tests/test_image_processing_common.py [hunk 2]
```diff
@@ -282,6 +282,43 @@ def _assert_tensors_equivalence(self, tensor1, tensor2, atol=1e-1, rtol=1e-3, me
...
+            reference_value, value = reference_encoding[key], encoding[key]
+            if not (torch.is_tensor(reference_value) and torch.is_tensor(value)):
+                continue
```
Comment (zucchini-nlp) [inline tests/test_image_processing_common.py]: what type of non-tensor inputs we can get? IIRC some models on purpose keep inputs as `list` which we still can check

File: tests/test_image_processing_common.py [hunk 3]
```diff
@@ -282,6 +282,43 @@ def _assert_tensors_equivalence(self, tensor1, tensor2, atol=1e-1, rtol=1e-3, me
         torch.testing.assert_close(tensor1, tensor2, atol=atol, rtol=rtol)
         self.assertLessEqual(torch.mean(torch.abs(tensor1 - tensor2)).item(), mean_atol)
 
+    def _assert_masks_equivalence(self, mask1, mask2):
+        """Masks are discrete, so the backends must agree on both the dtype and every value."""
+        self.assertEqual(mask1.dtype, mask2.dtype)
+        self.assertTrue(torch.equal(mask1, mask2))
```
Comment (zucchini-nlp) [inline tests/test_image_processing_common.py]: imo this is not needed - `_assert_encodings_equivalence` will do the same no?

File: tests/models/smolvlm/test_image_processing_smolvlm.py [hunk 4]
```diff
@@ -340,9 +338,7 @@ def test_backends_equivalence_batched(self):
         for backend_name in backend_names[1:]:
             encoding = encodings[backend_name]
             self._assert_tensors_equivalence(reference.pixel_values, encoding.pixel_values, atol=3e-1)
-            self._assert_tensors_equivalence(
-                reference.pixel_attention_mask.float(), encoding.pixel_attention_mask.float()
-            )
+            self._assert_masks_equivalence(reference.pixel_attention_mask, encoding.pixel_attention_mask)
             self.assertEqual(reference.rows, encoding.rows)
```
Comment (zucchini-nlp) [inline tests/models/smolvlm/test_image_processing_smolvlm.py]: no need to override this test anymore? (same comment for other models)

File: src/transformers/models/fuyu/image_processing_pil_fuyu.py [hunk 5]
```diff
@@ -211,6 +211,7 @@ def _preprocess(
                 "image_unpadded_heights": image_unpadded_heights,
                 "image_unpadded_widths": image_unpadded_widths,
                 "image_scale_factors": image_scale_factors,
+                "image_sizes": original_image_sizes,
```
Comment (zucchini-nlp) [inline src/transformers/models/fuyu/image_processing_pil_fuyu.py:214]: niiice, i suppose it was left out since `processing_xxx` tests run only with fast image processor 😢 

(gotta add one processing test with `use_fast=False`, not on you though)

**Review by zucchini-nlp — COMMENTED** [review]
Thanks for making the test suite better ❤️ Left a couple comments, prob no override needed anymore

**Comment by zucchini-nlp** [comment]
Feel free to ignore tests if not caused by your PR, I will re-run beforee merging

File: tests/test_image_processing_common.py [hunk 6]
```diff
@@ -282,6 +282,43 @@ def _assert_tensors_equivalence(self, tensor1, tensor2, atol=1e-1, rtol=1e-3, me
...
+
+    def _assert_encodings_equivalence(self, reference_encoding, encoding, reference_backend, backend_name):
```
Comment (molbap) [inline tests/test_image_processing_common.py]: indeed that is a pretty cool test!

**Review by molbap — COMMENTED** [review]
Thanks, and thanks for explaining how you did it + that you verified all cases 🙏 
Agreed one test can likely be removed, other than that looks fine

File: tests/test_image_processing_common.py (hunk 2 above)
Comment (yupengtang) [reply tests/test_image_processing_common.py]: Good point, they are compared now instead of skipped. Lists and tuples are walked item by item (tensors inside them use the same float tolerance as top-level tensors) and anything else goes through `assertEqual`. Across the current `backends_equivalence` tests the non-tensor outputs are eomt `patch_offsets`, minicpmv4_6 `grids` / `num_patches_per_image`, mllama `num_tiles`, Fuyu's nested `images` and SmolVLM / Idefics3 `rows` / `cols`, and all of them match between backends.

File: tests/test_image_processing_common.py (hunk 3 above)
Comment (yupengtang) [reply tests/test_image_processing_common.py]: Removed, the overrides go through `_assert_encodings_equivalence` now.

File: tests/models/smolvlm/test_image_processing_smolvlm.py (hunk 4 above)
Comment (yupengtang) [reply tests/models/smolvlm/test_image_processing_smolvlm.py]: I tried dropping them: the SmolVLM and Idefics3 overrides are still needed for their inputs, since with the base test's image and resampling `pixel_values` are not close between backends (it fails every run). They resize a 100x150 image with BICUBIC and request `return_row_col_info=True`. So I kept those parts and replaced the per-key asserts with `_assert_encodings_equivalence`, which also picks up `rows` and `cols`. The Idefics2 batched override stays for batches where samples have different numbers of images, and the Fuyu one no longer needs its separate `images[0][0]` check.

Pushed in 2afd4d3ba1. All `backends_equivalence` tests: 190 passed, 63 skipped. SmolVLM, Idefics2, Idefics3 and Fuyu image processing tests: 95 passed, 17 skipped, and the batched ones passed 5 runs in a row (they pop random images). On the source from before this PR the updated tests still fail 7 of 8, on the `pixel_attention_mask` dtype and Fuyu's missing `image_sizes`.

Comment (zucchini-nlp) [reply tests/models/smolvlm/test_image_processing_smolvlm.py]: i didnt get why we can'tdelete override, because we compare num-row and cols? If yes, no need for that as longas we compare pixels

Also pixels across backends should be similar so if base tests fails, there has to be a reason why

Comment (yupengtang) [reply tests/models/smolvlm/test_image_processing_smolvlm.py]: You're right, sorry, my earlier answer was wrong. I had torchvision 0.26 locally, and there the torchvision backend falls back from LANCZOS to BICUBIC while PIL uses real LANCZOS, so the pixels drifted. With torchvision 0.27 (which is what the CI image installs) the base tests pass for both SmolVLM and Idefics3, so I removed both overrides in 98fc27c.

Checked on torchvision 0.27.1: the touched model tests give 95 passed / 17 skipped, and the backend tests passed 5 runs out of 5. Without the source fix the base tests still fail 7 of 8 on the mask dtype and the missing `image_sizes`, so they still cover the bug.

**Review by zucchini-nlp — APPROVED** [review]
oke, lets merge when CI passes
